Add flags to disable per-volume metric reporting

deniseyu opened the pull request:
By default, the Datadog agent's disk reporter wants to
  export metrics for every volume. Concourse workers have
  a LOT of volumes and we're not interested in this degree
  of cardinality.

Signed-off-by: Denise Yu <dyu@pivotal.io>

I know we normally try to be additive with changes to ops files, but I seriously don't believe that anyone wants this level of cardinality for disk usage. We're not concerned about disk usage on these volumes, only on the hosts.

Commit:
Add flags to disable per-volume metric reporting

* By default, the Datadog agent's disk reporter wants to
  export metrics for every volume. Concourse workers have
  a LOT of volumes and we're not interested in this degree
  of cardinality.

Signed-off-by: Denise Yu <dyu@pivotal.io>

cirocosta commented on cluster/operations/datadog.yml:
```diff
@@ -34,6 +34,24 @@
+              tag_by_filesystem: true
+              all_partitions: true
+              device_blacklist:
+              - /var/vcap/data/worker/work/volumes/*/**
```
Hey, 

I don't think this is right - currently, it raises the following err on datadog-agent `7.16`:

```
2020-01-02 16:48:56 UTC | CORE | WARN | (pkg/collector/python/datadog_agent.go:118 in LogMessage) | unknown | (disk.py:321) | /var/vcap/data/worker/work/volumes/*/** is not a valid regular expression and will be ignored
```

If I understood correctly, this is supposed to be a regular expression, rather than glob matching:

```yaml
## @param device_blacklist - list of regex - optional
## Instruct the check to not collect from matching devices.
##
## Character casing is ignored on Windows. For convenience, the regular
## expressions start matching from the beginning and therefore to match
## anywhere you must prepend `.*`. For exact matches append `$`.
##
## When conflicts arise, this will override `device_whitelist`.
#
# device_blacklist:
#   - /dev/sde
#   - [FJ]:
```


https://github.com/DataDog/integrations-core/blob/bdfb6eb0a1fbeab9dff06e5726ec8556fc40f6da/disk/datadog_checks/disk/data/conf.yaml.default#L65-L76

which seems to be the case indeed by looking at the way that this configuration value gets consumed:


```python
    def _device_blacklisted(self, device):
        if not device or self._device_blacklist is None:
            return False

        return not not self._device_blacklist.match(device)
```

where the thing being `match`ed against is indeed a regex:

```python
        self._device_blacklist = self._compile_valid_patterns(
            self._device_blacklist, extra_patterns=device_blacklist_extras
        )
```

```python
    def _compile_valid_patterns(self, patterns, casing=IGNORE_CASE, extra_patterns=None):
        valid_patterns = []

         # ....
    
        if valid_patterns:
            return re.compile('|'.join(valid_patterns), casing)
```

https://github.com/DataDog/integrations-core/blob/cae4e8a77f3539a091899b911fcd60db0c815bfb/disk/datadog_checks/disk/disk.py#L292-L294 

I guess we could go with something like 

```
^\/var\/vcap\/data\/worker\/work\/volumes\/.*
```

wdyt?

thanks!

deniseyu replied:
ahh interesting. I tried to copy a pattern I saw in an email from a Datadog support person. But let's try the regex and see what happens - you want to do it or want me to? I might not have access to prod systems at some point very soon because I already sent off my offboarding request.

cirocosta replied:
oh true hahah np! I can take it 😁  thanks!

cirocosta replied:
damn, tried quite a bit, many of the rules I tried just didn't really work - we still get those measurements with all of the fields included (tried with `mount_point_blacklist` too) 😞  I'll give it a try another time soon

cirocosta commented on cluster/operations/datadog.yml:
```diff
@@ -34,6 +34,24 @@
+              all_partitions: true
+              device_blacklist:
+              - /var/vcap/data/worker/work/volumes/*/**
+              excluded_filesystems:
```
Hey,

This seems to raise a warning log:

```
2020-01-02 16:48:56 UTC | CORE | WARN | (pkg/collector/python/datadog_agent.go:118 in LogMessage) |  | (disk.py:272) | `excluded_filesystems` is deprecated and will be removed in 6.9. Please use `file_system_blacklist` instead.
```

It seems like `excluded_filesystems` is deprecated:

```python
    def _compile_pattern_filters(self, instance):
        # Force exclusion of CDROM (iso9660)
        file_system_blacklist_extras = ['iso9660$']
        device_blacklist_extras = []
        mount_point_blacklist_extras = []


        deprecation_message = '`{old}` is deprecated and will be removed in 6.9. Please use `{new}` instead.'


        if 'excluded_filesystems' in instance:
            file_system_blacklist_extras.extend(
                '{}$'.format(pattern) for pattern in instance['excluded_filesystems'] if pattern
            )
            self.warning(deprecation_message.format(old='excluded_filesystems', new='file_system_blacklist'))
```

(https://github.com/DataDog/integrations-core/blob/cae4e8a77f3539a091899b911fcd60db0c815bfb/disk/datadog_checks/disk/disk.py#L260-L272)

and yeah, indeed, it's not a config that's shown in the default config, but, instead, only the new one:

```yaml
    ## @param file_system_blacklist - list of regex - optional
    ## Instruct the check to not collect from matching file systems.
    ##
    ## Character casing is ignored. For convenience, the regular expressions
    ## start matching from the beginning and therefore to match anywhere you
    ## must prepend `.*`. For exact matches append `$`.
    ##
    ## When conflicts arise, this will override `file_system_whitelist`.
    #
    # file_system_blacklist:
```

https://github.com/DataDog/integrations-core/blob/bdfb6eb0a1fbeab9dff06e5726ec8556fc40f6da/disk/datadog_checks/disk/data/conf.yaml.default#L40-L49

Wdyt of  getting that changed in follow up PR?

deniseyu replied:
I can remove it from this PR!

cirocosta commented:
Hey,

I was trying to get a better sense of how this all gets down to the `disk`
collector, so here's the "trace" (might be useful for us in the future).

---

First, by configuring `generate_disk_config` to `false` in the ops file, and
passing `disk_yaml_config`, the job takes care of placing all of the data under
`dd.disk_yaml_config` to the file that the `disk` collector from `dd-agent`
looks at:


```
  dd.generate_disk_config:
    default: yes
    description: Generate disk configuration, disk.yaml

  dd.disk_yaml_config:
    default: ""
    description: Disk integration YAML configuration
```

(see [job spec](https://github.com/DataDog/datadog-agent-boshrelease/blob/7e52342d4b4460055ad288b1c258119c8d6e5581/jobs/dd-agent/spec#L197))


```ruby
<% if p('dd.generate_disk_config') == true || p('dd.generate_disk_config') =~ (/(true|t|yes|y|1)$/i) %>
mkdir -p ${CONFD_DIR}/disk.d
cat <<EOF > "${CONFD_DIR}/disk.d/disk.yaml"
---
init_config:
instances:
  - use_mount: yes
    tag_by_filesystem: <%= p("dd.generate_disk_config_tag_by_filesystem", "yes") %>
    all_partitions: <%= p("dd.generate_disk_config_all_partitions", "yes") %>
    excluded_filesystems:
      - tracefs
EOF
<% elsif p('dd.disk_yaml_config') != "" %>
mkdir -p ${CONFD_DIR}/disk.d
cat <<EOF > "${CONFD_DIR}/disk.d/disk.yaml"
<%= p('dd.disk_yaml_config') %>
EOF
<% end %>
```

(see [template](https://github.com/DataDog/datadog-agent-boshrelease/blob/39d9becc0c6cbc724a14ffb39fb0d989f4a013af/jobs/dd-agent/templates/config/confd.sh.erb#L167-L186))

Which, in the case of this PR, makes it write down to the disk config
(`/var/vcap/data/jobs/dd-agent/config/conf.d/disk.d/disk.yaml`):


```
init_config:
instances:
  - use_mount: yes
    tag_by_filesystem: true
    all_partitions: true
    device_blacklist:
    - /var/vcap/data/worker/work/volumes/*/**
    excluded_filesystems:
      - tracefs
```


The configuration we pass here has some interesting properties:


1. we do not use the default per-device collection (we set `use_mount: true`)

```yaml
## @param use_mount - boolean - required
## Instruct the check to collect using mount points instead of volumes.
#
use_mount: false
```


2. we do not go with the default of not adding a tag that mentions the fs (thus,
   we tag each measurement w/ its corresponding filesystem) - seems like a nice
   to have (we set `tag_by_filesystem: true`)


```yaml
## @param tag_by_filesystem - boolean - optional
## Instruct the check to tag all disks with their file system e.g. filesystem:ntfs.
#
# tag_by_filesystem: false
```

3. we blacklist all the mounts under `/var/vcap/data/worker/work/*/**`, the
   place where `concourse worker` mounts each of the volumes that it creates.
   This is currently not right (that's not a valid regexp), but the idea
   remains.


```yaml
## @param device_blacklist - list of regex - optional
## Instruct the check to not collect from matching devices.
##
## Character casing is ignored on Windows. For convenience, the regular
## expressions start matching from the beginning and therefore to match
## anywhere you must prepend `.*`. For exact matches append `$`.
##
## When conflicts arise, this will override `device_whitelist`.
#
# device_blacklist:
#   - /dev/sde
#   - [FJ]:
```


https://github.com/DataDog/integrations-core/blob/cdf89624ebed4372753f6f6f167608eeb955747d/disk/datadog_checks/disk/data/conf.yaml.default#L1



To test this in practice, I went with the `docker` approach as documented in
their guides (https://docs.datadoghq.com/agent/docker/?tab=standard) with the
following args:

        DOCKER_CONTENT_TRUST=1 docker run -d --name dd-agent \
                -v $(realpath ./datadog.yaml):/etc/datadog-agent/datadog.yaml \
                -v $(realpath ./disk.d/config.yaml):/etc/datadog-agent/conf.d/disk.d/config.yaml \
                -v /var/run/docker.sock:/var/run/docker.sock:ro \
                -v /proc/:/host/proc/:ro \
                -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
                -e DD_API_KEY=$API_KEY \
                        datadog/agent:7


where `datadog.yaml` is

        tags:
          - environment:ciro-test


and `disk/config.yaml` is whatever we want (e.g., the config above).


With regards to the implementation of disk checking, here's where all goes on:

- https://github.com/DataDog/integrations-core/blob/cae4e8a77f3539a091899b911fcd60db0c815bfb/disk/datadog_checks/disk/disk.py#L55-L68

(despite the "agent" being in Go, the "checker" is in python)

cirocosta commented:
Hey,

It seems like we'll need to do similar work for network ifaces too (despite the number being an order of magnitude compared to disk mounts. it's still pretty big - at least one network iface per container):

<img width="974" alt="Screen Shot 2020-01-02 at 12 38 08 PM" src="https://user-images.githubusercontent.com/3574444/71682192-e6c2f900-2d5c-11ea-92e5-177456840c7a.png">

^ for a single host w/out many containers, we can see that we have ~45 values for `device` on a single host 

ps.: this value at least rotates every hour, assuming each of those containers are check containers (thus, generating more timeseries inside datadog - assuming their doing timeseries in a standard way)

I'll set up an issue for that too (might lower a lot the load on some folks' datadog accounts).

Commit:
operations: ignore bc mounts from disk metrics

this commit improves the blacklisting by leveraging
`mount_point_blacklist` rather than `device_blacklist` (1), and in that
rule, using a regexp (2).

1: during the filtering that takes place in the disk collector, despite
   us having specified `use_mount`, it looks at `mount_point_blacklist`
   for mount point filtering (which, in the case of
   `/var/.../volumes/...`, it's just a mountpoint).

2: the disk collector does not perform glob expansion, but instead, it
   uses regexes (w/ ignorecase set).

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

Commit:
operations: replace deprecated `excluded_filesystems` in dd-agent

the `excluded_filesystems` configuration has been deprecated from the
dd-agent disk collector[1].

this commit changes that to use the new configuration
(`file_system_blacklist`).

[1]: https://github.com/DataDog/integrations-core/blob/cae4e8a77f3539a091899b911fcd60db0c815bfb/disk/datadog_checks/disk/disk.py#L268-L272

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

cirocosta commented:
hey @deniseyu , I just update the configuration (see last two commits). 

wdyt?

it got us down to a good number (fixed size) of distinct vals:

<img width="970" alt="Screen Shot 2020-01-02 at 8 02 10 PM" src="https://user-images.githubusercontent.com/3574444/71726533-66a49e00-2e05-11ea-8bb6-f23aae7d34f6.png">

(^ that drop is from the moment I applied the config)

<img width="865" alt="Screen Shot 2020-01-03 at 8 40 18 AM" src="https://user-images.githubusercontent.com/3574444/71726539-6a382500-2e05-11ea-883c-e088a8441e45.png">

thanks!
